From 91ee255055d9c2d5df22b47482ca60e2084be8c9 Mon Sep 17 00:00:00 2001
From: Vipul Singh <imvipul@google.com>
Date: Sun, 10 May 2026 23:47:51 -0700
Subject: [PATCH 14/27] Revert^2 "Writing ApplicationInfoCache instead of
 multiple ApplicationInfos for nested RemoteViews"

This reverts commit 532a0a90b30a8045efdf1b0a84249a2352482607.

Reason for revert: Fixing original bug

Bug: 477278752
Merged-In: I9caf4aedc7d2ba1697e06a5e46cd605bac07cad0
Change-Id: I9caf4aedc7d2ba1697e06a5e46cd605bac07cad0
(cherry picked from commit 2137c24fb2d16addba712a6844a275dfd0349351)
---
 core/java/android/widget/RemoteViews.java | 99 ++++++++++++++++-------
 1 file changed, 68 insertions(+), 31 deletions(-)

diff --git a/core/java/android/widget/RemoteViews.java b/core/java/android/widget/RemoteViews.java
index 09a2fd1c5a51..baaaee43c352 100644
--- a/core/java/android/widget/RemoteViews.java
+++ b/core/java/android/widget/RemoteViews.java
@@ -380,6 +380,11 @@ public class RemoteViews implements Parcelable, Filter {
     @UnsupportedAppUsage
     public ApplicationInfo mApplication;
 
+    /**
+     * Id referencing the ApplicationInfo in ApplicationInfoCache that hosts the remote views.
+     */
+    private int mApplicationId = -1;
+
     /**
      * The resource ID of the layout file. (Added to the parcel)
      */
@@ -1264,39 +1269,64 @@ public class RemoteViews implements Parcelable, Filter {
      * {@link ApplicationInfo} instance is used throughout the RemoteViews.
      */
     private static class ApplicationInfoCache {
-        private final Map<Pair<String, Integer>, ApplicationInfo> mPackageUserToApplicationInfo;
+        private final List<ApplicationInfo> mCachedApplicationInfos;
 
         ApplicationInfoCache() {
-            mPackageUserToApplicationInfo = new ArrayMap<>();
+            mCachedApplicationInfos = new ArrayList<>();
+        }
+
+        ApplicationInfoCache(Parcel parcel) {
+            mCachedApplicationInfos = parcel.createTypedArrayList(ApplicationInfo.CREATOR);
         }
 
         /**
-         * Adds the {@link ApplicationInfo} to the cache if it's not present. Returns either the
-         * provided {@code applicationInfo} or a previously added value with the same package name
-         * and uid.
+         * Adds the {@link ApplicationInfo} to the cache if it's not present. Returns an int Id
+         * which can be used to retrieve the {@link ApplicationInfo} from the cache.
          */
-        @Nullable
-        ApplicationInfo getOrPut(@Nullable ApplicationInfo applicationInfo) {
-            Pair<String, Integer> key = getPackageUserKey(applicationInfo);
-            if (key == null) return null;
-            return mPackageUserToApplicationInfo.computeIfAbsent(key, ignored -> applicationInfo);
+        int getOrPut(@Nullable ApplicationInfo applicationInfo) {
+            if (applicationInfo == null) return -1;
+            int id = findId(applicationInfo);
+            if (id != -1) return id;
+            mCachedApplicationInfos.add(applicationInfo);
+            return mCachedApplicationInfos.size() - 1;
+        }
+
+        int findId(@NonNull ApplicationInfo applicationInfo) {
+            if (applicationInfo == null) return -1;
+            for (int i = 0; i < mCachedApplicationInfos.size(); i++) {
+                ApplicationInfo cachedApplicationInfo = mCachedApplicationInfos.get(i);
+                if (Objects.equals(cachedApplicationInfo.packageName, applicationInfo.packageName)
+                        && cachedApplicationInfo.uid == applicationInfo.uid) {
+                    return i;
+                }
+            }
+            return -1;
         }
 
         /** Puts the {@link ApplicationInfo} in the cache, replacing any previously stored value. */
-        void put(@Nullable ApplicationInfo applicationInfo) {
-            Pair<String, Integer> key = getPackageUserKey(applicationInfo);
-            if (key == null) return;
-            mPackageUserToApplicationInfo.put(key, applicationInfo);
+        int updateAndGetId(@Nullable ApplicationInfo applicationInfo) {
+            if (applicationInfo == null) return -1;
+            int id = findId(applicationInfo);
+            if (id != -1) {
+                mCachedApplicationInfos.set(id, applicationInfo);
+                return id;
+            } else {
+                mCachedApplicationInfos.add(applicationInfo);
+                return mCachedApplicationInfos.size() - 1;
+            }
         }
 
         /**
          * Returns the currently stored {@link ApplicationInfo} from the cache matching
          * {@code  applicationInfo}, or null if there wasn't any.
          */
-        @Nullable ApplicationInfo get(@Nullable ApplicationInfo applicationInfo) {
-            Pair<String, Integer> key = getPackageUserKey(applicationInfo);
-            if (key == null) return null;
-            return mPackageUserToApplicationInfo.get(key);
+        @Nullable ApplicationInfo get(int id) {
+            return (id < 0 || id >= mCachedApplicationInfos.size())
+                    ? null : mCachedApplicationInfos.get(id);
+        }
+
+        void writeToParcel(Parcel dest, int flags) {
+            dest.writeTypedList(mCachedApplicationInfos, flags);
         }
     }
 
@@ -6334,7 +6364,7 @@ public class RemoteViews implements Parcelable, Filter {
     protected RemoteViews(ApplicationInfo application, @LayoutRes int layoutId) {
         mApplication = application;
         mLayoutId = layoutId;
-        mApplicationInfoCache.put(application);
+        mApplicationId = mApplicationInfoCache.updateAndGetId(application);
     }
 
     private boolean hasMultipleLayouts() {
@@ -6384,6 +6414,7 @@ public class RemoteViews implements Parcelable, Filter {
                     "Both RemoteViews must share the same package and user");
         }
         mApplication = portrait.mApplication;
+        mApplicationId = mApplicationInfoCache.updateAndGetId(mApplication);
         mLayoutId = portrait.mLayoutId;
         mViewId = portrait.mViewId;
         mLightBackgroundLayoutId = portrait.mLightBackgroundLayoutId;
@@ -6437,6 +6468,7 @@ public class RemoteViews implements Parcelable, Filter {
 
         RemoteViews smallestView = findSmallestRemoteView();
         mApplication = smallestView.mApplication;
+        mApplicationId = mApplicationInfoCache.updateAndGetId(mApplication);
         mLayoutId = smallestView.mLayoutId;
         mViewId = smallestView.mViewId;
         mLightBackgroundLayoutId = smallestView.mLightBackgroundLayoutId;
@@ -6522,6 +6554,7 @@ public class RemoteViews implements Parcelable, Filter {
             mIsRoot = false;
         }
         mApplication = src.mApplication;
+        mApplicationId = mApplicationInfoCache.updateAndGetId(mApplication);
         mLayoutId = src.mLayoutId;
         mLightBackgroundLayoutId = src.mLightBackgroundLayoutId;
         mApplyFlags = src.mApplyFlags;
@@ -6599,13 +6632,15 @@ public class RemoteViews implements Parcelable, Filter {
             mBitmapCache = new BitmapCache(parcel);
             // Store the class cookies such that they are available when we clone this RemoteView.
             mClassCookies = parcel.copyClassCookies();
+            mApplicationInfoCache = new ApplicationInfoCache(parcel);
             mCollectionCache = new RemoteCollectionCache(parcel);
         } else {
             configureAsChild(rootData);
         }
 
         if (mode == MODE_NORMAL) {
-            mApplication = parcel.readTypedObject(ApplicationInfo.CREATOR);
+            mApplicationId = parcel.readInt();
+            mApplication = mApplicationInfoCache.get(mApplicationId);
             mIdealSize = parcel.readInt() == 0 ? null : SizeF.CREATOR.createFromParcel(parcel);
             mLayoutId = parcel.readInt();
             mViewId = parcel.readInt();
@@ -6627,6 +6662,7 @@ public class RemoteViews implements Parcelable, Filter {
             initializeSizedRemoteViews(remoteViews.iterator());
             RemoteViews smallestView = findSmallestRemoteView();
             mApplication = smallestView.mApplication;
+            mApplicationId = mApplicationInfoCache.updateAndGetId(mApplication);
             mLayoutId = smallestView.mLayoutId;
             mViewId = smallestView.mViewId;
             mLightBackgroundLayoutId = smallestView.mLightBackgroundLayoutId;
@@ -6636,6 +6672,7 @@ public class RemoteViews implements Parcelable, Filter {
             mPortrait =
                     new RemoteViews(parcel, getHierarchyRootData(), mLandscape.mApplication, depth);
             mApplication = mPortrait.mApplication;
+            mApplicationId = mApplicationInfoCache.updateAndGetId(mApplication);
             mLayoutId = mPortrait.mLayoutId;
             mViewId = mPortrait.mViewId;
             mLightBackgroundLayoutId = mPortrait.mLightBackgroundLayoutId;
@@ -6780,7 +6817,8 @@ public class RemoteViews implements Parcelable, Filter {
         // Before propagating down the tree, replace our application from the root application info
         // cache, to ensure the same instance is present throughout the hierarchy to allow for
         // squashing.
-        mApplication = mApplicationInfoCache.getOrPut(mApplication);
+        mApplicationId = mApplicationInfoCache.getOrPut(mApplication);
+        mApplication = mApplicationInfoCache.get(mApplicationId);
 
         HierarchyRootData rootData = getHierarchyRootData();
         if (hasSizedRemoteViews()) {
@@ -6807,7 +6845,8 @@ public class RemoteViews implements Parcelable, Filter {
         if (!mIsRoot) return;
         mBitmapCache = new BitmapCache();
         mApplicationInfoCache = new ApplicationInfoCache();
-        mApplication = mApplicationInfoCache.getOrPut(mApplication);
+        mApplicationId = mApplicationInfoCache.getOrPut(mApplication);
+        mApplication = mApplicationInfoCache.get(mApplicationId);
         configureDescendantsAsChildren();
     }
 
@@ -9032,7 +9071,7 @@ public class RemoteViews implements Parcelable, Filter {
 
     /** @hide */
     public void updateAppInfo(@NonNull ApplicationInfo info) {
-        ApplicationInfo existing = mApplicationInfoCache.get(info);
+        ApplicationInfo existing = mApplicationInfoCache.get(mApplicationInfoCache.findId(info));
         if (existing != null && !existing.sourceDir.equals(info.sourceDir)) {
             // Overlay paths are generated against a particular version of an application.
             // The overlays paths of a newly upgraded application are incompatible with the
@@ -9042,7 +9081,7 @@ public class RemoteViews implements Parcelable, Filter {
 
         // If we can update to the new AppInfo, put it in the cache and propagate the change
         // throughout the hierarchy.
-        mApplicationInfoCache.put(info);
+        mApplicationInfoCache.updateAndGetId(info);
         configureDescendantsAsChildren();
     }
 
@@ -9437,9 +9476,10 @@ public class RemoteViews implements Parcelable, Filter {
             // is shared by all children.
             if (mIsRoot) {
                 mBitmapCache.writeBitmapsToParcel(dest, flags);
+                mApplicationInfoCache.writeToParcel(dest, flags);
                 mCollectionCache.writeToParcel(dest, flags, intentsToIgnore);
             }
-            dest.writeTypedObject(mApplication, flags);
+            dest.writeInt(mApplicationId);
             if (mIsRoot || mIdealSize == null) {
                 dest.writeInt(0);
             } else {
@@ -9454,6 +9494,7 @@ public class RemoteViews implements Parcelable, Filter {
             dest.writeInt(MODE_HAS_SIZED_REMOTEVIEWS);
             if (mIsRoot) {
                 mBitmapCache.writeBitmapsToParcel(dest, flags);
+                mApplicationInfoCache.writeToParcel(dest, flags);
                 mCollectionCache.writeToParcel(dest, flags, intentsToIgnore);
             }
             dest.writeInt(mSizedRemoteViews.size());
@@ -9466,6 +9507,7 @@ public class RemoteViews implements Parcelable, Filter {
             // is shared by all children.
             if (mIsRoot) {
                 mBitmapCache.writeBitmapsToParcel(dest, flags);
+                mApplicationInfoCache.writeToParcel(dest, flags);
                 mCollectionCache.writeToParcel(dest, flags, intentsToIgnore);
             }
             mLandscape.writeToParcel(dest, flags);
@@ -10644,12 +10686,6 @@ public class RemoteViews implements Parcelable, Filter {
         return viewId;
     }
 
-    @Nullable
-    private static Pair<String, Integer> getPackageUserKey(@Nullable ApplicationInfo info) {
-        if (info == null || info.packageName ==  null) return null;
-        return Pair.create(info.packageName, info.uid);
-    }
-
     private HierarchyRootData getHierarchyRootData() {
         return new HierarchyRootData(mBitmapCache, mCollectionCache,
                 mApplicationInfoCache, mClassCookies);
@@ -10900,6 +10936,7 @@ public class RemoteViews implements Parcelable, Filter {
                 checkProtoResultNotNull(ref.mUid, "No uid");
                 rv.mApplication = context.getPackageManager().getApplicationInfoAsUser(
                         ref.mPackageName, /* flags= */ 0, UserHandle.getUserId(ref.mUid));
+                rv.mApplicationId = rv.mApplicationInfoCache.updateAndGetId(rv.mApplication);
                 appContext = rv.getContextForResourcesEnsuringCorrectCachedApkPaths(context);
                 appResources = appContext.getResources();
 
-- 
2.55.0

